File: BOJ/[1238] 파티/JFe/1238.py
import sys, heapq
from collections import defaultdict

def dijkstra(node):
    dist = [int(1e9) for _ in range(N + 1)]
    dist[node] = 0
    h = []
    heapq.heappush(h, (0, node))
    while h:
        value, node = heapq.heappop(h)
        if value < dist[node]: continue
        for next_node, cost in graph[node]:
            next_value = value + cost
            if next_value < dist[next_node]:
                dist[next_node] = next_value
                heapq.heappush(h, (next_value, next_node))
    return dist

N, M, X = map(int, sys.stdin.readline().split())
road_info = [[int(x) for x in sys.stdin.readline().split()] for _ in range(M)]
graph = defaultdict(list)
# graph 입력
for a, b, t in road_info:
    graph[a].append([b, t])
# 목적지 X에서 각 마을로 갈 수 있는 최단 거리 계산
x_dist = dijkstra(X)
result = 0
# 각 마을에서 목적지 X로 갈 수 있는 최단 거리 구하고 + 목적지 X -> i 마을까지 최단 거리 더해서
# 가장 오래 걸린 시간 출력
for i in range(1, N + 1):
    time = dijkstra(i)[X] + x_dist[i]
    result = max(result, time)
print(result)

File: BOJ/[1238] 파티/KingDonggyu/1238.js
function dequeue(queue) {
  const [node, value, idx] = queue.reduce(
    (acc, [node, value], i) => {
      return acc[1] > value ? [node, value, i] : acc;
    },
    [null, Infinity, 0]
  );
  queue.splice(idx, 1);
  return [node, value];
}

function dijkstra(N, graph, start) {
  const queue = [[start, 0]];
  const minTime = Array(N + 1).fill(Infinity);
  minTime[start] = 0;
  while (queue.length) {
    const [node, time] = dequeue(queue);
    graph[node].forEach(([nextNode, newTime]) => {
      if (minTime[nextNode] <= time + newTime) {
        return;
      }
      minTime[nextNode] = time + newTime;
      queue.push([nextNode, minTime[nextNode]]);
    });
  }
  return minTime;
}

function solution(N, X, graph) {
  const reversedGraph = Array.from(Array(N + 1), () => []);
  graph.forEach((edges, src) => {
    edges.forEach(([dst, time]) => {
      reversedGraph[dst].push([src, time]);
    })
  })
  const goTimes = dijkstra(N, reversedGraph, X);
  const backTimes = dijkstra(N, graph, X);
  let answer = 0;
  for (let i = 1; i < N + 1; i++) {
    answer = Math.max(answer, goTimes[i] + backTimes[i]);
  }
  return answer;
}

const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const input = [];
let N, M, X;

rl.on('line', (line) => {
  input.push(line);
  if (input.length === 1) {
    [N, M, X] = input[0].split(' ').map((c) => +c);
    return;
  }
  if (input.length === M + 1) {
    rl.close();
  }
}).on('close', () => {
  const graph = Array.from(Array(N + 1), () => []);
  input.slice(1).forEach((s) => {
    const [src, dst, time] = s.split(' ');
    graph[+src].push([+dst, +time]);
  });
  console.log(solution(N, X, graph));
  process.exit();
});
